lib/amazon_ssa_support/amazon_ssa_extract.rb

```diff
@@ -0,0 +1,174 @@
+end
+
+require 'aws-sdk'
+require_relative 'instance_metadata'
+require_relative 'evm_heartbeat'
+require_relative 'evm_queue_extractor'
+require_relative 'rolling_s3_outputter'
+
+class LogFormatter < Log4r::Formatter
+  @@prog          = File.basename(__FILE__, ".*")
+  @@extractor_id  = ''
+  def format(event)
+    return event.data.chomp + "\n" if event.level == Log4r::COPY
+    
+    "#{Log4r::LNAMES[event.level]} [#{datetime}" +
+    (@@extractor_id.nil? ? "" : " #{@@extractor_id}") +
+    "] -- #{@@prog}: " +
+    (event.data.kind_of?(String) ? event.data : event.data.inspect) + "\n"
+  end
+  
+  def self.extractor_id=(val)
+    @@extractor_id  = val
+  end
+
+  private
+
+  def datetime
+    time = Time.now.utc
+    time.strftime("%Y-%m-%dT%H:%M:%S.") << "%06d" % time.usec
+  end
+end
+
+LOG_LEVELS  = {
+  'DEBUG'       => Log4r::DEBUG,
+  'INFO'        => Log4r::INFO,
+  'WARN'        => Log4r::WARN,
+  'ERROR'       => Log4r::ERROR,
+  'FATAL'       => Log4r::FATAL,
+  Log4r::DEBUG  => "DEBUG",
+  Log4r::INFO   => "INFO",
+  Log4r::WARN   => "WARN",
+  Log4r::ERROR  => "ERROR",
+  Log4r::FATAL  => "FATAL"
+}
+
+cmdName       = File.basename($0)
+log_level_str = nil
+max_log_size  = nil
+log_to_stderr = false
+
+#
+# Process command line args.
+#
+OptionParser.new do |opts|
+  opts.banner = "Usage: #{cmdName} [options]"
+
+  opts.on('--log-to-stderr', "Log to stderr in addition to the log file")  do
+    log_to_stderr = true
+  end
+  opts.on('-l', '--loglevel ARG', "The log level: DEBUG|INFO|WARN|ERROR|FATAL")  do |ll|
+    raise OptionParser::ParseError.new("Unrecognized log level: #{ll}") if !(/DEBUG|INFO|WARN|ERROR|FATAL/i =~ ll)
+    log_level_str = ll.upcase
+  end
+  opts.on('--max-log-size ARG', "Roll to S3 after log file exceeds this size")  do |mls|
+    n = max_log_size.to_i
+    raise OptionParser::ParseError.new("Max log size must be >= 256") if n < 256
+    max_log_size = n
+  end
+
+  begin
+    opts.parse!(ARGV)
+  rescue OptionParser::ParseError => perror
+    $stderr.puts cmdName + ": " + perror.to_s
+    $stderr.puts
+    $stderr.puts opts.to_s
+    exit 1
+  end
+end
+
+begin
+
+  #
+  # Get info and args associated with this extractor instance.
+  #
+  im = Amazon::SSA::InstanceMetadata.new
+  extractor_id = im.metadata('instance-id')
+  aws_args = eval(im.userdata)
+  aws_args[:extractor_id] = extractor_id
+  aws_args[:evm_bucket] = "evm-prototype"
+  reg = aws_args[:region] || 'us-west-2'
+  
+  aws_args[:ec2] = Aws::EC2::Resource.new(region: reg)
+  aws_args[:sqs] = Aws::SQS::Resource.new(region: reg)
+  aws_args[:s3]  = Aws::S3::Resource.new(region: reg)
+  
+  #
+  # Logging args.
+  #
+  aws_args[:log_level]  = log_level_str unless log_level_str.nil? # command line override
+  log_level        = LOG_LEVELS[aws_args[:log_level]]
+  max_log_size      ||= aws_args[:max_log_size] || 1024 * 256
+  logFile          = File.join('/opt/miq/log', 'extract.log')
```

Comment:
@hsong-rh yes.  I actually will be using the gem from within manageiq-providers-amazon as well so I can use the EvmQueue methods.